### PR TITLE
Add test simulating bad connection between pageserver and compute

### PR DESCRIPTION
**Pull request by save-buffer**
## Problem
We have a funny 3-day timeout for connections between the compute and pageserver. We want to get rid of it, so to do that we need to make sure the compute is resilient to connection failures.  

Closes: https://github.com/neondatabase/neon/issues/5518

## Summary of changes
This test makes the pageserver randomly drop the connection if the failpoint is enabled, and ensures we can keep querying the pageserver. 

## Checklist before requesting a review

- [x] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

File: pageserver/src/page_service.rs
Comment (koivunej) [inline pageserver/src/page_service.rs:958]: Regardless of my recent https://github.com/neondatabase/neon/pull/5696 I think we are going to be flushing for these connections. Ugh, I so wish we had removed this anonymous inner class alike PostgresQueryHandler already because it is making everything more difficult to see. Also, solving that good will be a difficult one.

Perhaps we could in addition to returning an error also cancel the cancellation token? That would give us fast exit. If we are always using `task_mgr::shutdown_token` for shutdown, then it will be fine to cancel. We most likely are using it.

File: test_runner/regress/test_bad_connection.py
Comment (jcsp) [inline test_runner/regress/test_bad_connection.py:8]: nit: can you make the test names a bit more specific?  Something like `test_compute_pageserver_connection_stress`, so that it's a bit more obvious what kind of  connections we're talking about.

File: pageserver/src/page_service.rs
Comment (jcsp) [inline pageserver/src/page_service.rs:207]: I think we can avoid hardcoding `CONNECTION_FAILURE_PROBABILITY` & get it in the failpoint configuration that's passed in when enabling it:

```
The format of a single action is [p%][cnt*]task[(arg)]. p% is the expected probability that the action is triggered, and cnt* is the max times the action can be triggered. The supported values of task are:
```

https://docs.rs/failpoints/latest/failpoints/fn.cfg.html

File: pageserver/src/page_service.rs
Comment (jcsp) [inline pageserver/src/page_service.rs:954]: Does returning an error from here result in a disconnection?  Looking at `PostgresBackend::process_message`, it seems like errors from `process_query` just send an error response and proceed to continue servicing requests on the connection.

File: pageserver/src/page_service.rs
Comment (save-buffer) [reply pageserver/src/page_service.rs:207]: Nice, I missed that somehow. Changed it

File: test_runner/regress/test_bad_connection.py
Comment (save-buffer) [reply test_runner/regress/test_bad_connection.py:8]: Changed

File: pageserver/src/page_service.rs
Comment (save-buffer) [reply pageserver/src/page_service.rs:954]: Good point, I was checking the logs initially and I saw that the failpoint was getting hit, but now looking at it again it seems that it's only the first failpoint that gets hit. Once a connection is established it never goes away. I'll need to find another way

File: pageserver/src/page_service.rs
Comment (save-buffer) [reply pageserver/src/page_service.rs:958]: Looks like cancellation token isn't making it actually disconnect.

Comment (koivunej) [reply pageserver/src/page_service.rs:958]: Even if you cancel the token within the failpoint? It is very difficult to eyeball this. Perhaps we do need to actually kill the err the connection, but at the same time I would not expect you to be able to return a connection reset error here if nothing useful was done to it.

(I wrote #5733 in the meantime about this big picture problem with this abstraction.) 

We could have a failpointable TcpStream wrapper which is later composed to `tokio_io_timeout`, which would drop the actual connection as perhaps the layers of abstraction is just too much on this side.

Comment (save-buffer) [reply pageserver/src/page_service.rs]: OK i just decided to have a failpoint-configurable average timeout that follows an exponential distribution to test it. This does seem to create a lot of connection errors on the compute side, and does degrade performance (as expected. It usually executes 1500 times with a good connection and now executes 600 times with a bad one), but the workload does still seem to execute successfully

File: pageserver/src/page_service.rs
Comment (save-buffer) [reply pageserver/src/page_service.rs]: See my other comment to Joonas about the configurable timeout strategy now.

File: pageserver/src/page_service.rs
Comment (jcsp) [inline pageserver/src/page_service.rs:233]: Using a timeout to cause disconnections is highly likely to only exercise disconnections in between requests: the latency to serve a getpage request is in the microseconds, so this test setting a 15ms timeout is probably just going to wait until the client is idle, and then disconnect.

As always, the litmus test for this kind of thing is to point wireshark at it and check what kind of behavior you're really getting, but from reading the code it seems like this isn't going to inject the kind of mid-request disconnects that are most stressful.

Comment (jcsp) [reply pageserver/src/page_service.rs:233]: I think the original version of the code attempting to inject disconnections between a request and a response was on the right track: the only problem was that `process_query` couldn't return a value that caused a disconnect.

Perhaps you could add a `cfg(testing)` variant to `QueryError` that `PostgresBackend::process_message` understands to mean "disconnect now".

Then that, combined with the idle timeout case you've got here, would cover both disconnections between requests (the relatively easy case) and disconnections during requests (the slightly harder request, perhaps).

Comment (save-buffer) [reply pageserver/src/page_service.rs:233]: > Perhaps you could add a cfg(testing) variant to QueryError that PostgresBackend::process_message understands to mean "disconnect now".

That's a good idea, not sure why I didn't think of it earlier. Just added this (though I didn't make it `#[cfg(feature = testing)]` because it wasn't compiling and I couldn't really figure out why).

File: pageserver/src/page_service.rs
Comment (jcsp) [inline pageserver/src/page_service.rs:235]: Please can you remove this comment "timeout should be lower..." before merging -- it's no longer relevant now that the default timeout is reduced to 4205

**Review by jcsp — APPROVED** [review]
LGTM -- when this merges, please flag in #team-compute that the pageserver timeout has been reduced, so that folks are prepared in case we see any issues.

**Pull request merged**
